Add the notion of Component

thiblahute opened the pull request:
Components is basically a set of target on which we bind a user
option to enable or disable it.

This is basically used to auto generate options to enable and disable
certain parts of a build.

In GStreamer we have many plugins and the user/packager often need
to be able to explicitely say what they want to build and what not. Currently
we would need to add **many** option to allow that and have `if option(xxx)` all
around. The notion of components allows us to do it all very simply.

There are several ways to create a component:

* Creating explicitely in the meson.build file:

```meson
component = component('componentname',
    targets: [],
    enabled: false, # disabled by default, meaning the option is
                    # --enable-componentname instead of the default
                    # --disable-componentname
    dependencies: [another_component], # @componentname is disabled
                                       # if @another_component is disabled
)
```

* Specifying it as a string or reusing previsouly creating one creating a target

```meson
target = library(..., components: ['componentname'])
```

* Specifying it as a string or reusing previsouly creating one for all targets in a subdir

It means that all targets created inside those subdirectories will be contained in
the specified components. You can do so with:

```meson
subdir('somesubdir', components: 'componentname')
```

ignatenkobrain commented on mesonbuild/interpreter.py:
```diff
@@ -155,6 +156,23 @@ def prepend_method(self, args, kwargs):
 
 
+
+class Component(MutableInterpreterObject):
```
__[flake8]__
* [E303] too many blank lines (3)

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/optinterpreter.py:
```diff
@@ -72,10 +72,16 @@ def __init__(self, subproject, command_line_options):
+
+    @staticmethod
+    def get_command_line_options(subproject, command_line_options):
+        sbprfeix = subproject + ':'
```
__[flake8]__
* [F841] local variable 'sbprfeix' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/optinterpreter.py:
```diff
@@ -72,10 +72,16 @@ def __init__(self, subproject, command_line_options):
-            if self.subproject != '': # Strip the beginning.
-                if not o.startswith(self.sbprefix):
+            if subproject != '': # Strip the beginning.
+                if not o.startswith(sbprefix):
```
__[flake8]__
* [F821] undefined name 'sbprefix'

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on run_unittests.py:
```diff
@@ -124,8 +124,12 @@ def tearDown(self):
-                                               env=os.environ.copy())
+        try:
+            self.output += subprocess.check_output(command, stderr=subprocess.STDOUT,
+                                                env=os.environ.copy())
```
__[flake8]__
* [E128] continuation line under-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->

jpakkane commented:
Could you write a use case on how this is meant to be used and what specific problem it is solving. Link to specific use cases in GStreamer if possible. Something like:

> We need to build <X> in a way to enable <Y>, and because current features are not sufficient we need to work around this by doing <Z> and this makes things easier by enabling <W>.

Thanks,

thiblahute commented:
> Could you write a use case on how this is meant to be used and what specific problem it is solving. Link to specific use cases in GStreamer if possible. Something like:

>> We need to build X in a way to enable Y, and because current features are not sufficient we need to work around this by doing Z and this makes things easier by enabling W.

Updated the description to explain a bit our use case in GStreamer.

thiblahute commented:
Ping?

jpakkane commented:
Sorry, this has been postponed by the releases. Will look into this over the weekend.

jpakkane commented:
The basic idea is nice and something like this is generally useful. There are a bunch of questions that need to be thought through first:

 - probably the options would be `on`, `off` and `auto`, which means building if dependencies are found, how would those be expressed
 - should components get their own keyword or would putting them to `dependencies` work just as well?
 - when turning a subdir into a component what happens to build targets that are not libraries (executables, custom targets)? They should probably be ignored but could there be cases where that causes problems?

I'd really like to keep the core as minimal as possible. If we could achieve the same UX without adding a new primitive objec type, it would be a lot better. I need to think about this.

thiblahute commented:
> The basic idea is nice and something like this is generally useful. There are a bunch of questions that need to be thought through first:
> 
> * probably the options would be on, off and auto, which means building if dependencies are found, how would those be expressed

Well the option is either `-Denable-component-name` or `-Ddisable-component-name` (thanks to `enabled` arg when building a component) if nothing specified by the user, it means `auto` why do you think we need something different?

> * should components get their own keyword or would putting them to dependencies work just as well?

Well, letting the user work with a component object is pretty usefull. For example you may  want to generate documenation per component (which is what I am doing in my GStreamer hotdoc branch) making the whole thing elegant fmpov. 

> * when turning a subdir into a component what happens to build targets that are not libraries (executables, custom targets)? They should probably be ignored but could there be cases where that causes problems?

Currently they are also added to the component, meaning anyone outside the component that depends on it is not built, but if it is not what the user want, he should not have turned that subdir into a component. I thought about that and decided it was the best thing to do as having those special cased would actually be less intuitive fmpov.

> I'd really like to keep the core as minimal as possible. If we could achieve the same UX without adding a new primitive objec type, it would be a lot better. I need to think about this.

I think this primitive object adds a very useful concept that can actually solve several things at once, I am not sure you can have anything like that without adding a new primitive object but I am happy to hear sugestions.

jpakkane commented:
> Well the option is either -Denable-component-name or -Ddisable-component-name (thanks to enabled arg when building a component) if nothing specified by the user, it means auto why do you think we need something different?

Let's look at this from the user point of view. At some point there is going to be user code that looks like this:

    if some_component.was_found()
      <something>
    endif

It needs to check the object rather than deciphering via project options whether something was enabled or not (because it can't really tell). This implies that the call to `subdir` needs to look something like this:

    subdir('foo',
      enabling_option : 'build_foo', # on, off or auto
      needed_deps : [dep1, dep2], # deps that must have been found for auto mode
      component_varname : 'some_component', # set to a not_found component if dir was not entered
    )

This enters the subdir only if option value is `true` or if it is `auto` and all the deps listed are found. It also does the fallback.

This also splits up setups between two directories, which is not a great experience.

> I am not sure you can have anything like that without adding a new primitive object but I am happy to hear sugestions.

If we ignore executables, then components are roughly equivalent to declared dependencies. What we are looking for is a better UX for joining them with togglable options. As an example one thing we could do is that instead of having the logic in the subdir call we could have something like this in the actual subdir:

    check_buildability_and_return_if_false(option_name : 'build_foo',
      needed_deps : [dep1, dep2],
      varname : 'foo_dep',
    )
    <build the thing>
    foo_dep = declare_dependency(...)

If the requirements are not met (as above), this sets the given variable to a non-found dependency and returns. This avoids needing to wrap the entire build definition in an `if/else`.

tp-m commented:
I wouldn't ignore executables or custom targets. Some use cases I would like to handle with this is to switch off building executables, tests or docs for example (in addition to the plugins scenario).

My first impression is that a new object type would be best, for three reasons:
 * discoverability, of what is an extremely handy core feature that most projects are likely to want to make use of in some form.
 * extensibility (chances are we'll think of more tweaks/features/scenarios in future)
 * clarity and focus (for lack of a better word) - ideally we really don't want to add multiple new kwargs to lots of other functions

The main thing I'm not sure how to make work nicely yet from a user point of view is the auto-thing. Often this will be a 1:N mapping to one or more dependency(), but sometimes there will have to be a more elaborate check, which should probably only be done if the component was not disabled.

Just on the side as an additional point to consider, there's also the scenario where we have components for all the plugins (base scenario discussed above) but then also a higher-level component for 'all plugins with external dependencies', and perhaps also orthogonal components for 'decoding' and 'encoding'; various codec plugins may only include certain sources for compilation if decoding is not disabled or encoding is not disabled. This might just add confusion for now though, so feel free to ignore it. I haven't seen anything yet that would preclude this use case.

ignatenkobrain commented on run_unittests.py:
```diff
@@ -1493,6 +1493,36 @@ def test_run_installed(self):
+        test_data = [
+            (['-Dhello1=disabled'], 'disabled', 'enabled', 'disabled', 'enabled', 'enabled'),
+            (['-Dhello1=disabled', '-Dhello2=disabled'],
+              'disabled', 'disabled', 'disabled', 'enabled', 'enabled'),
```
__[flake8]__
* [E127] continuation line over-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on run_unittests.py:
```diff
@@ -1493,6 +1493,36 @@ def test_run_installed(self):
+            hello1 = os.path.join(self.builddir, 'hello1')
+            hello2 = os.path.join(self.builddir, 'hello2')
+            hello3 = os.path.join(self.builddir, 'hello3')
+            hello4 = os.path.join(self.builddir, 'hello4')
```
__[flake8]__
* [F841] local variable 'hello4' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->

thiblahute commented:
> > Well the option is either -Denable-component-name or
> > -Ddisable-component-name (thanks to enabled arg when building a
> > component) if nothing specified by the user, it means auto why do
> > you think we need something different? Let's look at this from the
> > user point of view. At some point there is going to be user code
> > that looks like this:
>
>     if some_component.was_found()
>       <something>
>     endif

Possible now.

> It needs to check the object rather than deciphering via project
> options whether something was enabled or not (because it can't really
> tell). This implies that the call to subdir needs to look something
> like this:
>
>     subdir('foo',
>       enabling_option : 'build_foo', # on, off or auto
>       needed_deps : [dep1, dep2], # deps that must have been found for auto mode
>       component_varname : 'some_component', # set to a not_found component if dir was not entered
>     )
>
> This enters the subdir only if option value is true or if it is auto
> and all the deps listed are found. It also does the fallback.

This does not sound so nice tbh, I would rather let the user do proper
checks in the given directory, meaning that the component is enabled
or auto, or pass the right dependencies when creating the component.


> This also splits up setups between two directories, which is not a
> great experience.
>
>> I am not sure you can have anything like that without adding a new
>> primitive object but I am happy to hear suggestions.
>
> If we ignore executables, then components are roughly equivalent to
> declared dependencies. What we are looking for is a better UX for
> joining them with togglable options. As an example one thing we could
> do is that instead of having the logic in the subdir call we could
> have something like this in the actual subdir:
>
>     check_buildability_and_return_if_false(option_name : 'build_foo',
>       needed_deps : [dep1, dep2],
>       varname : 'foo_dep',
>     )
>
>     <build the thing>
>     foo_dep = declare\_dependency(...)
>
> If the requirements are not met (as above), this sets the given
> variable to a non-found dependency and returns. This avoids needing to
> wrap the entire build definition in an if/else.

Well, this proposal is about easily adding options for build targets
in general and manipulating those options in a user friendly way.
This also includes a way to know what targets belong to a component
and I am sure this proposal can cover new things we have not thought
about yet.

I do not see how that idea covers what I am trying to solve even though
it also solves part of it.

I started porting GStreamer plugins, and you can see an example in
[GStreamer Base] (fmpov this is nice to use).

  [GStreamer Base]: https://github.com/thiblahute/gst-plugins-base/commit/c328babdee6df10ff743c555f3039303c1f541b5

> Just on the side as an additional point to consider, there's also the
> scenario where we have components for all the plugins (base scenario
> discussed above) but then also a higher-level component for 'all
> plugins with external dependencies', and perhaps also orthogonal
> components for 'decoding' and 'encoding'; various codec plugins may
> only include certain sources for compilation if decoding is not
> disabled or encoding is not disabled. This might just add confusion
> for now though, so feel free to ignore it. I haven't seen anything yet
> that would preclude this use case.

I do not see any problem about that with this proposal :-)

---

I pushed a rebased enhanced branch that also contains documentation, you can find
the rendered documentation [here](https://thiblahute.github.io/jpakkane.github.io/Reference-manual.html#component)

jpakkane commented:
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

thiblahute commented:
Ping again, it would be nice to make a decision about that one, we really need something like that in GStreamer and it has been ~6 months already :)

tp-m commented:
I wanted @nirbheek to have a look at it as well and have been pestering him about it. I would very much like this in for GStreamer 1.14 as well, I'm sure we'll get it done in time (i.e. before Sept/Oct).

leio commented:
Being able to disable components is sort of a must-have for my gstreamer packaging in Gentoo, so +1 in getting this in time for gst 1.14 usage, as this seems to help achieve that.
We also need a way to easily disable all of those in ext/ and sys/ (without maintaining a list manually, preferably), and then selectively enable only certain ones, due to splitting them at source package level.
Currently with autotools version of gst-plugins-* build system we achieve that via some creative AG_GST_CHECK_FEATURE based sed on configure.ac; good if we could achieve this in a smarter way with meson too.

jpakkane commented:
> Being able to disable components is sort of a must-have for my gstreamer packaging in Gentoo

What specific requirements do you have? The simplest solution (which I think is what gst already does) is to not build plugins whose dependencies are not found. What are the things you need in addition to that?

tp-m commented:
I think the following functionality, is required:
 * `enable-foo=yes`: enable and error out if required deps are not available (the erroring out may have to be punted to the user in some cases)
 * `disable-foo`: disable and ideally don't even check deps
 * `enable-foo=auto`: enable if required deps are available, otherwise not.

Ideally there should also be a disable-reason available somewhere for later printing, but that's not so important for starters.

leio commented:
Not building plugins whose dependencies are not found is the worst thing that could happen from a packaging point of view. I need to be able to be in control what is built and what isn't, in particular with those that do have external dependencies. The package manager has to know what dependencies are there in the end result and force them accordingly too.
This is source level packaging splitting, as we can't do it after building like binary distributions do one source package and many binary packages - we have to do it at the source level.

We have a big package for all the gstreamer tarballs (gst-plugins-base, gst-plugins-good, gst-plugins-ugly, gst-plugins-bad), which have to build all the non-plugin stuff, all the gst/ (without extra deps) plugins and all the helper libraries (gst-libs/). And then we have separate packages for each of the plugins in sys/ or ext/ subdirs. These packages need to only build and install that plugin subdirectory, and be able to link against system helper libraries of the same tarball, instead of in-tree. With autotools based gstreamer we just call make/make install inside those dirs after full configure run (this is where I'm mostly hoping meson will help remove the only real downside of the source package split - huge configure runtime for each split package). And the helper library stuff is solved with autotools via some more sedding - https://gitweb.gentoo.org/repo/gentoo.git/tree/eclass/gstreamer.eclass#n169

That whole gstreamer.eclass file is then how we achieve all that without any need for doing anything special in the split package ebuilds themselves right now. We only need to pass some extra variables if the package name doesn't match the subdirectory name or the subdirectory name doesn't match the plugin name, iirc. Plus there are certain cases where we actually build ext/ or sys/ plugins as part of the main package, primarily in case of gst-plugins-base. This currently simply works by the eclass adding --disable-* for all the ext/ sys/ things, and then the ebuild appending --enable-* for them if appropriate and the last one takes effect as that's how autoconf works, https://gitweb.gentoo.org/repo/gentoo.git/tree/media-libs/gst-plugins-base/gst-plugins-base-1.10.4.ebuild#n48

Here's a random example of a split plugin package, where we need to override the plugin name/builddir due to package and plugin name difference:
https://gitweb.gentoo.org/repo/gentoo.git/tree/media-plugins/gst-plugins-gdkpixbuf/gst-plugins-gdkpixbuf-1.10.4.ebuild - and that's all we need to do together with expressing the extra deps, and all just works with autotools when combined with the "hacks" in the eclass.

thiblahute commented:
Rebased + fixed conflicts + renamed tests to `155 component` now that 154 is used.

nirbheek commented:
So this is generally nice IMO, but it's a bit unclear what the current state of the feature is, so I'll try to summarize how I would like it to look:

 * `component()` as documented in the Reference Manual in this PR — LGTM, except:
    * `is_enabled()` should be just `enabled()`, for consistency with `found()`
    * `list_targets()` should be `get_targets()` or just `targets()`. I like the latter more because this method takes no arguments.
 * `subdir()` kwarg `components:`
    * Should only take component objects, not strings
    * Parser simply should not `subdir()` if *any* of the components are disabled or not found
    * Optional components will not be known to `subdir()`, instead they will be used via `component.found()` inside the build file(s) in the subdirectory
* `build_target()` and `custom_target()`—same as `subdir()`.

Is there anything I'm missing?

nirbheek commented:
One more thing, we should have clear output in the configure log when a component is disabled, when we don't subdir() because a component is not enabled, and same for any target(s). This might also be a good time to implement printing of a summary at the end of all components that were enabled or disabled.

thiblahute commented:
> So this is generally nice IMO, but it's a bit unclear what the current state of the feature is, so I'll try to summarize how I would like it to look:
> - [x] component() as documented in the Reference Manual in this PR — LGTM, except:
>     * [x] is_enabled() should be just enabled(), for consistency with found()
>     * [x] list_targets() should be get_targets() or just targets(). I like the latter more because this method takes no arguments.
> - [ ] subdir() kwarg components:
>   - [x] Should only take component objects, not strings
>   - [x] Parser simply should not subdir() if any of the components are disabled or not found
>   - [ ] Optional components will not be known to subdir(), instead they will be used via component.found() inside the build file(s) in the subdirectory

I am not sure what you mean by that, especially because of your previous point :-).
Having `components` in the `subdir` is pretty useful to me.

> - [x] build_target() and custom_target()—same as subdir().

thiblahute commented:
> One more thing, we should have clear output in the configure log when a component is disabled, when we don't subdir() because a component is not enabled, and same for any target(s). This might also be a good time to implement printing of a summary at the end of all components that were enabled or disabled.

Well it is the case, example:

```
 INSERT  thiblahute  …  test cases  common  155 components  1  rm -R build; mkdir build ;meson build/                                                                                  ST 51   components 
The Meson build system
Version: 0.42.0.dev1
Source dir: /home/thiblahute/devel/gstreamer/gst-build/meson/test cases/common/155 components
Build dir: /home/thiblahute/devel/gstreamer/gst-build/meson/test cases/common/155 components/build
Build type: native build
Project name: component
Native C compiler: ccache cc (gcc 7.1.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Component hello1 : auto
Component hello3 : disabled
Component hello2 : auto
Component subcomponent : auto
Component subsubcomponent : auto
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Dependency nowaythisdepexists found: NO
Component withdep1 : auto
Build targets in project: 6
 INSERT  thiblahute  …  test cases  common  155 components  1  rm -R build; mkdir build ;meson build/ -Dsubsubcomponent=disabled                                                       ST 51   components 
The Meson build system
Version: 0.42.0.dev1
Source dir: /home/thiblahute/devel/gstreamer/gst-build/meson/test cases/common/155 components
Build dir: /home/thiblahute/devel/gstreamer/gst-build/meson/test cases/common/155 components/build
Build type: native build
Project name: component
Native C compiler: ccache cc (gcc 7.1.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Component hello1 : auto
Component hello3 : disabled
Component hello2 : auto
Component subcomponent : auto
Component subsubcomponent : disabled
Subdir:  subdir/subsubdir disabled by component: subsubcomponent
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Dependency nowaythisdepexists found: NO
Component withdep1 : auto
Build targets in project: 5
```

thiblahute commented:
Just added `component.dependencies()`

jpakkane commented:
We should probably have a meeting about this at Guadec.

thiblahute commented:
> We should probably have a meeting about this at Guadec.

Sadly I am not attending this year.

nirbheek commented:
@thiblahute 
> Well it is the case, example:

I meant having a summary at the end of configure similar to how we have on Autotools configure. This would be subproject-aware so you don't need to scroll up 300 lines to figure out whether or not some component was enabled or not.

> I am not sure what you mean by that, especially because of your previous point :-).
> Having components in the subdir is pretty useful to me.

I meant `subdir()` should not have a mode in which it tries to do something funky about components that are optional for the code in that subdir. The build file(s) inside that subdir should use `component.found()`. For example, we shouldn't try to put in complicated logic like "`subdir()` if component A or component B were found, but not component C". It should be simple: "`subdir()` if these components are found, if not, don't".

> Sadly I am not attending this year.

I'll post a summary on IRC of what we discussed, and you can tell us what you think of it?

nirbheek commented on mesonbuild/interpreter.py:
```diff
@@ -152,13 +152,15 @@ def prepend_method(self, args, kwargs):
         self.interpreter = interpreter
         self.option_name = option_name
         self.targets = set()
+        self.force_disable_reason = force_disable_reason
```
I actually think we should do this by writing a new `dependency()` type that does a bunch of compiler checks under the hood. There is demand for something like this, as we saw on IRC the other day with Edward. For example:

`cc.find_library()` + `cc.has_header()` or `cc.has_header_symbol()`

The advantage of this would be that the dependency object can know the include path(s) to use as well, so you don't have to pass them manually.

Is there some other use-case you were thinking of when you wrote this?

thiblahute replied:
> Is there some other use-case you were thinking of when you wrote this?

This was my main use case indeed.

tp-m commented:
I think we should ignore the summary printing aspect for now, it's really orthogonal to this PR and can always be added later. I'd rather focus on making sure we get the conceptual issues and user-facing APIs right :)

tp-m commented:
1.  As discussed with @thiblahute on IRC, there's one thing that currently doesn't work quite right yet IMHO, and that's the typical GStreamer external plugin use case, e.g.
```
alsa_dep = dependency('alsa', required : false)
shared_module('gstalsa', ..., components: component('alsa', dependencies: alsa_dep))
```
What will happen is that if someone specifies `--disable-alsa` we will still do the alsa check. I think this is very unexpected and it will confuse users.

The problem here is that we have a bit of a circular dependency: the component needs to know the deps and the dep needs to know about the component in order to skip the check if disabled.

A possible solution would be (h/t @thiblahute) to add a dependency automatically to a component if the component is specified on the dependency, i.e. our alsa plugin definition would then look like this:
```
alsa_comp = component('alsa')
alsa_dep = dependency('alsa', required: false, components: alsa_comp)
shared_module('gstalsa', ..., components: alsa_comp)
```

We could also make a special `subdir_component()` as a special thing at the very beginning of a subdir meson.build file just like `projects()` is now for the top-level meson.build file, and then assume everything else in the file is covered by the switch. That would just hide the mechanics though, and I think the explicit thing is not too bad at all.


2. Another thing: We need a meson command line arg to switch any 'auto' default for components to 'disable' or 'enable' (for packagers/buildbots), so that nothing will be built unless explicitly enabled, or it will error out if anything is missing that's not explicitly disabled.

ignatenkobrain commented on mesonbuild/interpreter.py:
```diff
@@ -151,6 +151,103 @@ def prepend_method(self, args, kwargs):
+        value = self.interpreter.environment.coredata.user_options[self.option_name].value
+        if value == 'enabled' and not dep.found():
+            raise InterpreterException("Component %s is enabled but the following"
+                                        " dependencies are not present: %s" % (
```
__[flake8]__
* [E127] continuation line over-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/interpreter.py:
```diff
@@ -2137,6 +2352,36 @@ def func_library(self, node, args, kwargs):
+
+        name = args[0]
+        deps = mesonlib.flatten(kwargs.get('dependencies', []))
+        default_value = kwargs.get( 'default',
```
__[flake8]__
* [E201] whitespace after '('

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/interpreter.py:
```diff
@@ -2137,6 +2352,36 @@ def func_library(self, node, args, kwargs):
+        name = args[0]
+        deps = mesonlib.flatten(kwargs.get('dependencies', []))
+        default_value = kwargs.get( 'default',
+                                   self.coredata.builtins['components_default_value'].value)
```
__[flake8]__
* [E128] continuation line under-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->

thiblahute commented:
@tp-m Implemented both 1) and 2) :-)

tp-m commented on mesonbuild/coredata.py:
```diff
@@ -323,6 +330,7 @@ def get_builtin_option_default(optname):
     'layout':          [UserComboOption, 'Build directory layout.', ['mirror', 'flat'], 'mirror'],
     'default_library': [UserComboOption, 'Default library type.', ['shared', 'static'], 'shared'],
     'backend':         [UserComboOption, 'Backend to use.', backendlist, 'ninja'],
+    'components_default_value':  [ComponentOption, 'Value to use by default for component options.', 'auto'],
```
Shouldn't it be more like: "Enable/Disable all components that default to autodetection"?

thiblahute replied:
Well, it is exactly what id does, you do `components_default_value=['enabled', 'disabled']` and it will make sure that all components that default to `auto` will default to that value.

tp-m replied:
What I was trying to say was that to me the description sounds like it will force this value on *all* components, not just the autodetect ones :)

thiblahute replied:
hah, right right, misinterpreted you, indeed! :-)

thiblahute replied:
Done.

ignatenkobrain commented on mesonbuild/interpreter.py:
```diff
@@ -977,9 +1099,98 @@ def find_library_method(self, args, kwargs):
+                disabled = True
+
+        if disabled:
+            return DependencyHolder(dependencies.ExternalDependency(name, self.environment,
```
__[flake8]__
* [F821] undefined name 'name'

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/interpreter.py:
```diff
@@ -151,6 +156,126 @@ def prepend_method(self, args, kwargs):
+        if dep not in self.dependencies:
+            if not dep.found() and self.func_enabled():
+                mlog.log('component', mlog.bold(self.name), 'is now disabled by:',
+                        mlog.red(dep.name))
```
__[flake8]__
* [E128] continuation line under-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->

jpakkane commented:
The overall design seems nice, but there are a bunch of details that need clarifying.

 - first of all the name _component_ seems a bit off somehow for this feature, since it is more of a "toggle switch combiner", but I can't think of a better name
 - what should be the return value of `executable` if the component is not enabled
 - this implementation creates new options inside Meson files which is explicitly against current practice, could we tie them to existing options instead
 - this would permit people to have an on/off switch for things that do not support auto (dunno if this is actually useful) with a boolean option or a combo option with `true/false/auto` (we could probably create a shortcut "autobool" option type for this) and have a single option affect many different components

thiblahute commented:
> The overall design seems nice, but there are a bunch of details that need clarifying.
> 
> first of all the name component seems a bit off somehow for this feature, since it is more of a "toggle switch combiner", but I can't think of a better name

Right, I could not either :-)

> what should be the return value of executable if the component is not enabled

It returns a `target` object, that won't be built as it is disabled by the component.

> this implementation creates new options inside Meson files which is explicitly against current practice, could we tie them to existing options instead

Defining them again manually is unnecessary extra work fmpov, what do you suggest exactly?

> this would permit people to have an on/off switch for things that do not support auto (dunno if this is actually useful) with a boolean option or a combo option with true/false/auto (we could probably create a shortcut "autobool" option type for this) and have a single option affect many different components

You can already have a component depending on other components, so have a single option affecting several components is already possible, also you can set the default value of a component, we could probably extend them to disallow `auto`, I can't see any use case for it yet though.

jpakkane commented:
> It returns a target object, that won't be built as it is disabled by the component.

But this is only the beginning. It raises a bunch of questions such as:

 - if you use the executable target in a `test`, what should happen?
 - if you try to use a library in a `link_with`, what should happen?
 - if you declare a dependency with it, what should happen, what should happen when you try to use it in a target
 - what should happen if you try to use it in a e.g. pkg-config creation

ignatenkobrain commented on mesonbuild/modules/pkgconfig.py:
```diff
@@ -134,6 +134,13 @@ def generate(self, state, args, kwargs):
+        for l in libs + priv_libs:
+            if isinstance(l, (build.SharedLibrary, build.StaticLibrary)):
+                if not l.should_build():
+                    mlog.log("Not building %s as depending library %s is not being built" %(
```
__[flake8]__
* [E225] missing whitespace around operator

<!-- Comment by [SideCI](https://sideci.com) -->

thiblahute commented:
> It returns a target object, that won't be built as it is disabled by the component.
> But this is only the beginning. It raises a bunch of questions such as:

Well, the answers are quite obvious to me, but thanks for asking them as they were unanswered in the implementation!

> if you use the executable target in a test, what should happen?

Test is not added and thus disabled. (Now implemented and tested)

> if you try to use a library in a link_with, what should happen?

The target is disabled as it is missing a dependency. (Now implemented and tested)

> if you declare a dependency with it, what should happen, what should happen when you try to use it in a target

It should not be declared as it has a missing dependency.  (Now implemented and tested)

> what should happen if you try to use it in a e.g. pkg-config creation

Same thing, not created  (Now implemented and tested)

ignatenkobrain commented on run_project_tests.py:
```diff
@@ -242,6 +242,10 @@ def validate_install(srcdir, installdir, compiler):
         if fname not in expected and not fname.endswith('.pdb') and compiler == 'cl':
             ret_msg += 'Extra file {0} found.\n'.format(fname)
+
+    if ret_message:
```
__[flake8]__
* [F821] undefined name 'ret_message'

<!-- Comment by [SideCI](https://sideci.com) -->

jpakkane commented:
Since the full thing seems to be problematic, how about starting with something simpler:

 - add the true/false/auto option
 - add a component object with the given functionality
 - make the component usable as a toggle in `subdir`

Would this be useful on its own? Maybe with an additional kwarg to `subdir` that allows it to set some variable to a given value as a fallback.

nirbheek commented:
I had an idea about this: since we can't define options on-the-fly from inside `meson.build`, how about we change the implementation such that components are defined in `meson_options.txt`?

With that change, instead of adding dependencies to components, you'd add components to dependencies (via strings):

```meson
dependency('foo', components: ['comp1', 'comp2'])
```

And the same with targets (which is already implemented).

You would be able to use `get_component('compname')` to get a component object on which you can call `.targets()` and `.enabled()`. This object can also be used instead of strings in kwargs. Strings are allowed for convenience because it would get tedious to repeatedly type `get_component()` for every single component (which can be ~100 per project, in the gstreamer case).

With this we don't have to define components twice (once in `meson_options.txt` for the option and once in `meson.build` for the object itself).

ignatenkobrain commented on run_project_tests.py:
```diff
@@ -215,11 +215,15 @@ def validate_install(srcdir, installdir, compiler):
     expected = {}
     ret_msg = ''
     # Generate list of expected files
+    allow_extra_files=False
```
__[flake8]__
* [E225] missing whitespace around operator

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on run_project_tests.py:
```diff
@@ -215,11 +215,15 @@ def validate_install(srcdir, installdir, compiler):
     elif os.path.exists(info_file):
         with open(info_file) as f:
             for line in f:
+                if  line.strip('\n') == '++':
```
__[flake8]__
* [E271] multiple spaces after keyword

<!-- Comment by [SideCI](https://sideci.com) -->

thiblahute commented:
> Since the full thing seems to be problematic, how about starting with something simpler:

Why do you say it is problematic? It involves little details to handle, but the whole idea of that feature is to handle it once and for all, avoiding meson user to have to take care themselves of the implication of disabling a part of the build. I missed some of them, clearly, but I believe this branch already handles *many* of them, and I do not understand what that statement comes from. If you talk about the issues in the test, it was due to details in the test themselves, mostly related to windows.

Also how does your proposal change implication of the concept? I do not see.

@nirbheek I did not understand having components expose option from the meson.build files was such an issue, why do you think it is a blocking point? If it is a no go, your proposal sounds good to me, but please explain why you think it is a problem.

thiblahute commented:
Ping, again.

nirbheek commented:
> Why do you say it is problematic?

Because of the restriction of not being able to define options implicitly when a component is declared in a build file.

> I did not understand having components expose option from the meson.build files was such an issue, why do you think it is a blocking point? If it is a no go, your proposal sounds good to me, but please explain why you think it is a problem.

I think we should be able to define options from inside build files, but Jussi disagrees (and I understand his reasons). That's why I proposed defining components in `meson_options.txt` because otherwise you have to define options twice, and that quickly becomes annoying.

jpakkane commented:
A component requires (at least) two different things:

1. an option of the correct type
2. a dependency

The former are defined only in option files and the latter only in build files. The latter can not be a string for a dependency name because it might be provided by one of many dependencies, especially if we add support for something like `dep = dependency('foobar', 'foobar12', 'some-other-name-for-foobar').

Thus all the information needed to specify when and how a component is defined is only available at runtime. As an example you could have this:

    d = dependency('foo', required : false)
    if something #such as d not found, specialty platform, some user option etc
      d = dependency('bar', required : false)
    endif

    c = component(option_name : 'mycomp',
      dependencies : d)

If we were to define components elsewhere, the `dependencies` value there would have to be a string rather than an object and it would need to be expanded at some point (it's not even clear when this should be, because dependencies can come and go at any time during a Meson run). This is not good because then you need to duplicate the dependency detection logic there which is not nice.

tp-m commented:
@jpakkane: note that we also want a pattern like
```
foo_dep = dependency('foo', components: ['plugin-foo''])
```
because otherwise there's no way to prevent meson checking for `foo` if the component is disabled (doing that is just very very bad user experience, a system that doesn't prevent that isn't fit for purpose imo).

A dependency is also not mandatory, e.g. you could just have a `subdir()` call being guarded by a component, or maybe executable targets.

jpakkane commented:
How about we go in a slightly different direction and call this a `dependency_group` instead? Then we could do something like:

    dg = dependency_group(option_name : 'somegroup', # an option with tristate values
      dependencies : ['foo', 'bar', dependency_object, dependencygroup_object])

If the option is `false`, the output is roughly the same as for `d = dependency('nonexisting', required : false)`.

If the option is `true`, all specified dependencies (and groups) must be found.

If the value is `auto`, then it checks for all listed dependencies. If they are found, then the return value is the same as if the value had been `true`. If any one of them is not found, then the return value is the same as if the option had been `false`.

thiblahute commented:
Well, functionally it is exactly the same thing to me, but requiring the user to register an option in the `meson_option.txt` file for it iiuc.

`option_group` is not a bad name either, not sure I prefer it though :-)

jpakkane commented:
Is there anything that the proposed  scheme does not do? Any requirements it does not fullfill?

We probably need to add something like:

    subdir('foo',
      if_found : some_dependency_or_group)

to make it nicer to use.

thiblahute commented:
> Is there anything that the proposed scheme does not do? Any requirements it does not fullfill?

Well it pretty much means duplicate entries about the `dependency_group`, once in the meson_option.txt and once in the meson.build files.

Also what does it bring compared to the component proposal?

jpakkane commented:
I pondered more about the disabling of dependents when things are missing. Here's a conceptual imagining of the issue. This is probably not directly applicable but hopefully moves us closer to a solution.

The two problems here is that we would like to disable things (such as test executables, other build targets etc) that depend on some target that is disabled as easily as possible but also to make it as clear as possible to the user _why_ something got disabled. Printing log messages as we go along is not a good UX because we already print too much stuff and things get lost.

For the first one suppose we had a specific type of an object, call it a `disabler`. It has very simple semantics. Whenever it is passed as an argument to a function (at least build target, test, etc) the function does nothing and its return value is the same disabler object. This would solve the former problem.

The latter problem could be solved by writing a standalone log file, say `meson-logs/disables.txt` whose contents look like this:

    subdir/fooexe disabled because of a disabling object that came from otherdir/lib/foolib
    ....

This would permit us to not print any disabling information on the basic log (which is useless 99% of the time) while also allowing very fast and precise debugging on those cases where it does go wrong.

jpakkane commented:
This is getting old in the tooth and the proper solution for this would probably use the disabler in some way. Closing this to reduce clutter on the tracker. Thanks.
